### ISSUE TITLE
Replacing reverse engineered Inception with official Inception architecture

### ISSUE THREAD

**Issue by bernardopires** [Russell91/TensorBox#18]
Hi Russel,

sorry that I'm creating yet another issue. I've been tinkering with your code and I think it would be great if it was using the official architecture instead of reverse engineered version. I've been trying to replace it with https://github.com/tensorflow/models/tree/master/inception, which is actually inception v3, so it also would be an 'upgrade' and possibly give even better performance.

I've started by replacing the call to `googlenet_load.model` to `inception_v3` (see https://github.com/tensorflow/models/blob/master/inception/inception/slim/inception_model.py). Since it's v3, we should use `mixed_8x8x2048b` instead of `mixed5b`, which is the layer right before the average pooling layer. The number of output filters is 2048 instead of 1024. 

Previously on inception v2, the output of `mixed5b` was exactly a  `Shape(batch_size, grid_height, grid_width, 1024)`. This is not the case for inception v3, which is returning a shape smaller for the height and width dimensions. What kind of correction would I have to do here? Inception v2 reduces the input size by a factor of 32 exactly but it seems like Inception v3 changes the shape slightly due to the used patch sizes.

Thanks!
Bernardo

**Comment by bernardopires** [comment]
I got it to run and while the training loss goes down, the test loss stabilizes quite early. The network didn't really learn anything. Here's an example image from tensorboard: http://i.imgur.com/gP89dom.png

And the training/test curves: http://i.imgur.com/Shcgs9S.png

**Comment by Russell91** [comment]
Interesting, do you have the source for the inception v3 model? I've gotten pretty good results on kitti with inception v1 (albeit with some attention mechanism), and it would be interesting to try out your changes.

**Comment by bernardopires** [comment]
What do you mean the source for the inception v3 model? Doesn't https://github.com/tensorflow/models/tree/master/inception have everything you're looking for?

**Comment by Russell91** [comment]
Yea, I mean your modified Tensorbox repository that has already hacked in the inception v3 model to the googlenet_load.

**Comment by bernardopires** [comment]
It seems like it's working now! It's still early on training but the bounding boxes are starting to make sense: http://i.imgur.com/bGvZfVf.png

My code is not in the best of shapes right now as I'm only interested in overfeat but basically all the changes are in here
```
from inception.inception_model import inference
def build_overfeat_forward(H, x, phase):
    input_mean = 117.
    x -= input_mean

    k = H['arch']['num_classes']
    dense_layer_num_output = [k, 4]
    features_dim = 2048
    with tf.variable_scope('inception_v3') as scope:
      W = [
          tf.get_variable('softmax/weights_{}'.format(i), initializer=tf.truncated_normal([features_dim, dense_layer_num_output[i]], stddev=0.001))
          for i in range(2)
      ]

      B = [
          tf.get_variable('softmax/biases_{}'.format(i), initializer=tf.random_normal([dense_layer_num_output[i]], stddev=0.001))
          for i in range(2)
      ]

    logits, endpoints = inference(images=x, num_classes=2, for_training=True, restore_logits=True)
    tf.get_variable_scope().reuse_variables()
    mixed5b = endpoints['mixed_8x8x2048b'] # right before avgpool layer
    mixed5b = tf.reshape(mixed5b, [H['arch']['batch_size'] * H['arch']['grid_width'] * H['arch']['grid_height'], features_dim])
    grid_size = H['arch']['grid_width'] * H['arch']['grid_height']
    pred_logits = tf.reshape(tf.nn.xw_plus_b(mixed5b, W[0], B[0],
                                             name=phase+'/logits_0'), 
                             [H['arch']['batch_size'] * grid_size, H['arch']['num_classes']])
    pred_confidences = tf.nn.softmax(pred_logits)
    pred_boxes = tf.reshape(tf.nn.xw_plus_b(mixed5b, W[1], B[1],
                                            name=phase+'/logits_1'), 
                            [H['arch']['batch_size'] * grid_size, 1, 4]) * 100
    return pred_boxes, pred_logits, pred_confidences
```
If you want I can polish the code, add support for the LSTMs and send a PR.

**Comment by bernardopires** [comment]
Oh this part is also very important:
```
    # load only the necessary layers from inception
    vars_to_load = []
    for var in tf.all_variables():
        if var.name.startswith(("conv0", "conv1", "conv2", "conv3", "conv4", "mixed_35x35", "mixed_17x17", "mixed_8x8")):
            vars_to_load.append(var)
   restore_inception = tf.train.Saver(vars_to_load)
   with tf.Session(config=config) as sess:
         if len(weights_str) > 0:
            print('Restoring from: %s' % weights_str)
            saver.restore(sess, weights_str)
            starting_step = int(weights_str.split('/')[-1].split('-')[-1])
        else:
            print('No starting point given, loading Inception v3')
            restore_inception.restore(sess, '%s/data/model.ckpt-157585' % os.path.dirname(os.path.realpath(__file__)))
            starting_step = 0
            print('Loaded Inception v3')
```

**Comment by Russell91** [comment]
Hey @bernardopires, been a little busy, but do you have any numbers on the brainwash dataset with the inceptionv3 model for comparison?
